src/backend/commands/analyze.c

```diff
@@ -2625,66 +2546,98 @@ acquire_sample_rows_dispatcher(Relation onerel, bool inh, int elevel,
 	 * Read the result set from each segment. Gather the sample rows *rows,
 	 * and sum up the summary rows for grand 'totalrows' and 'totaldeadrows'.
 	 */
-	funcRetValues = (char **) palloc0(funcTupleDesc->natts * sizeof(char *));
+	funcRetValues = (Datum *) palloc0(funcTupleDesc->natts * sizeof(Datum));
 	funcRetNulls = (bool *) palloc(funcTupleDesc->natts * sizeof(bool));
-	values = (char **) palloc0(relDesc->natts * sizeof(char *));
+	dvalues = (Datum *) palloc0(relDesc->natts * sizeof(Datum));
+	dnulls = (bool *) palloc(relDesc->natts * sizeof(bool));
 	sampleTuples = 0;
 	*totalrows = 0;
 	*totaldeadrows = 0;
-	for (int resultno = 0; resultno < cdb_pgresults.numResults; resultno++)
+
+	if (!portal->holdStore)
 	{
-		struct pg_result *pgresult = cdb_pgresults.pg_results[resultno];
-		bool		got_summary = false;
+		elog(ERROR, "No tuples received.");
+	}
+
+	slot = MakeSingleTupleTableSlot(queryDesc->tupDesc);
+	while (tuplestore_gettupleslot(portal->holdStore, true, false, slot))
+	{
+		TupleDesc	typeinfo = slot->tts_tupleDescriptor;
+		int			natts = typeinfo->natts;
+		Datum		attr;
+		bool		isnull;
 		double		this_totalrows = 0;
 		double		this_totaldeadrows = 0;
 
-		if (PQresultStatus(pgresult) != PGRES_TUPLES_OK)
+		/* There should be only one attribute with OID RECORDOID. */
```

Comment:
I wonder what are you doing here (and all the way down)? Why don't you use `create_memtuple_binding()` and `memtuple_deform()` but rather implement something strange with casting a memory tuple to the heap one?

Reply:
It is based on [record_out](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/utils/adt/rowtypes.c#L318) function that is [called](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/access/common/printtup.c#L487) from [printtup](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/access/common/printtup.c#L303) for record type.

> (gdb) bt
#0  printtup (slot=0x5598bfe31700, self=0x5598bfd2cf58) at printtup.c:487
#1  0x00005598bdda79e4 in ExecutePlan (estate=0x5598bfe30ad8, planstate=0x5598bfe31118, operation=CMD_SELECT, sendTuples=1 '\001', numberTuples=0,
    direction=ForwardScanDirection, dest=0x5598bfd2cf58) at execMain.c:3054
#2  0x00005598bdda3d32 in standard_ExecutorRun (queryDesc=0x5598bfdcaa58, direction=ForwardScanDirection, count=0) at execMain.c:983
#3  0x00005598bdda38ef in ExecutorRun (queryDesc=0x5598bfdcaa58, direction=ForwardScanDirection, count=0) at execMain.c:849
#4  0x00005598be00c1a5 in PortalRunSelect (portal=0x5598bfddcfd8, forward=1 '\001', count=0, dest=0x5598bfd2cf58) at pquery.c:1151
#5  0x00005598be00bda9 in PortalRun (portal=0x5598bfddcfd8, count=9223372036854775807, isTopLevel=1 '\001', dest=0x5598bfd2cf58, altdest=0x5598bfd2cf58,
    completionTag=0x7ffcdf57cb80 "") at pquery.c:992
#6  0x00005598be003784 in exec_simple_query (query_string=0x5598bfd2b978 "select pg_catalog.gp_acquire_sample_rows(16385, 10000, 'f');") at postgres.c:1820
#7  0x00005598be008dae in PostgresMain (argc=1, argv=0x5598bfd3d9d8, dbname=0x5598bfd0b178 "adb", username=0x5598bfd0b158 "gpadmin") at postgres.c:5242
#8  0x00005598bdf68367 in BackendRun (port=0x5598bfd3c4d0) at postmaster.c:4811
#9  0x00005598bdf6790a in BackendStartup (port=0x5598bfd3c4d0) at postmaster.c:4468
#10 0x00005598bdf6334a in ServerLoop () at postmaster.c:1948
#11 0x00005598bdf626f4 in PostmasterMain (argc=6, argv=0x5598bfd09010) at postmaster.c:1518
#12 0x00005598bde4adb3 in main (argc=6, argv=0x5598bfd09010) at main.c:245
(gdb) p thisState->finfo
$37 = {fn_addr = 0x5598be0e7a13 <record_out>, fn_oid = 2291, fn_nargs = 1, fn_strict = 1 '\001', fn_retset = 0 '\000', fn_stats = 2 '\002', fn_extra = 0x0,
  fn_mcxt = 0x5598bfd87098, fn_expr = 0x0}
(gdb) 


There were problems with using create_memtuple_binding due to record type so I have decided to stick to the  [record_out](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/utils/adt/rowtypes.c#L318) approach.

Reply:
But what was the problem with memory tuples? Why can't you simply deform them?

Reply:
I've looked at the code with more attention and understand the problem. As we dispatch `select pg_catalog.gp_acquire_sample_rows(...)` query we return a memory tuple with a record attribute type in result (so you check `RECORDOID` and number of attributes for this). Then you want to find a way to unpack the record and use `lookup_rowtype_tupdesc()` to get tuple's real description. But I still wonder why do you manually construct a heap tuple from the memory one to deform it?

Reply:
This approach to parse record using temporary HeapTuple control structure with heap_deform_tuple call seems to be very popular. For example, [here](https://github.com/arenadata/gpdb/blob/adb-6.x/contrib/hstore/hstore_io.c#L830-L843), [here](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/utils/adt/jsonfuncs.c#L2464-L2477), [here](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/utils/adt/jsonfuncs.c#L2810-L2821), and [here](https://github.com/arenadata/gpdb/blob/adb-6.x/src/test/regress/regress.c#L779-L790).